

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>txsscanlib.report &mdash; Txsscan 0.9a1 documentation</title>
    
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.9a1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="Txsscan 0.9a1 documentation" href="../../index.html" />
    <link rel="up" title="Module code" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../index.html">Txsscan 0.9a1 documentation</a> &raquo;</li>
          <li><a href="../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for txsscanlib.report</h1><div class="highlight"><pre>
<span class="c"># -*- coding: utf-8 -*-</span>

<span class="c">#===============================================================================</span>
<span class="c"># Created on Nov 29, 2012</span>
<span class="c"># </span>
<span class="c"># @author: bneron</span>
<span class="c"># @contact: user_email</span>
<span class="c"># @organization: organization_name</span>
<span class="c"># @license: license</span>
<span class="c">#===============================================================================</span>


<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="n">_log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s">&#39;txsscan.&#39;</span> <span class="o">+</span> <span class="n">__name__</span><span class="p">)</span>


<span class="kn">import</span> <span class="nn">abc</span>


<div class="viewcode-block" id="HMMReport"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.HMMReport">[docs]</a><span class="k">class</span> <span class="nc">HMMReport</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    handle HMM report. extract a synthetic report from the raw hmmer output</span>
<span class="sd">    this class is an abstract class. there is 2 implementation of this abstract class</span>
<span class="sd">    depending if the genome baes is ordered or not.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">__metaclass__</span> <span class="o">=</span> <span class="n">abc</span><span class="o">.</span><span class="n">ABCMeta</span>

<div class="viewcode-block" id="HMMReport.__init__"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.HMMReport.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">hmmer_output</span><span class="p">,</span> <span class="n">hmmer_err</span><span class="p">,</span> <span class="n">hmmer_returncode</span><span class="p">,</span> <span class="n">cfg</span> <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param gene: the gene corresponding to this profile</span>
<span class="sd">        :type gene: :class:`txsscanlib.gene.Gene` object</span>
<span class="sd">        :param hmmer_output: The path to hmmer output file</span>
<span class="sd">        :type hmmer_output: string</span>
<span class="sd">        :param hmmer_err: The path to the hmmer error file</span>
<span class="sd">        :type hmmer_err: string</span>
<span class="sd">        :param hmmer_returncode: the return code of the hmmer</span>
<span class="sd">        :type hmmer_returncode: int</span>
<span class="sd">        :param cfg: the configuration </span>
<span class="sd">        :type cfg: :class:`txsscanlib.config.Config` object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hmmer_raw_out</span> <span class="o">=</span> <span class="n">hmmer_output</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hmmer_err</span> <span class="o">=</span> <span class="n">hmmer_err</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hmm_returncode</span> <span class="o">=</span> <span class="n">hmmer_returncode</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_hmm_extracted</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_extract_out</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hits</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span> <span class="o">=</span> <span class="n">cfg</span>
</div>
    <span class="nd">@abc.abstractmethod</span>
<div class="viewcode-block" id="HMMReport.extract"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.HMMReport.extract">[docs]</a>    <span class="k">def</span> <span class="nf">extract</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse the output file of hmmer and produced a new synthetic report file.</span>
<span class="sd">        containing selected and sorted hits. </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span>
</div>
    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s">&quot;# gene: </span><span class="si">%s</span><span class="s"> extract from </span><span class="si">%s</span><span class="s"> hmm output</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_hmmer_raw_out</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">&quot;# profile length= </span><span class="si">%d</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">profile</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">&quot;# i_evalue threshold= </span><span class="si">%f</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">i_evalue_sel</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">&quot;# coverage threshold= </span><span class="si">%f</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">coverage_profile</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">&quot;# hit_id replicon_name position_hit gene_name gene_system i_eval score coverage</span><span class="se">\n</span><span class="s">&quot;</span>
        <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">:</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">s</span>

<div class="viewcode-block" id="HMMReport.save_extract"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.HMMReport.save_extract">[docs]</a>    <span class="k">def</span> <span class="nf">save_extract</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        write the string representation of the extact report in a file.</span>
<span class="sd">        the name of this file is the concatenation of the gene and the res_extract_suffix from config</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">extract_out_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">res_extract_suffix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_extract_out</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">res_search_dir</span><span class="p">,</span> <span class="n">extract_out_name</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_extract_out</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">_file</span><span class="p">:</span>
            <span class="n">_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>
</div>
<div class="viewcode-block" id="HMMReport.best_hit"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.HMMReport.best_hit">[docs]</a>    <span class="k">def</span> <span class="nf">best_hit</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        return the best hit when multiple hits</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">None</span>
</div></div>
<div class="viewcode-block" id="UnOrderedHMMReport"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.UnOrderedHMMReport">[docs]</a><span class="k">class</span> <span class="nc">UnOrderedHMMReport</span><span class="p">(</span><span class="n">HMMReport</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    handle HMM report. extract a synthetic report from the raw hmmer output</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="UnOrderedHMMReport.extract"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.UnOrderedHMMReport.extract">[docs]</a>    <span class="k">def</span> <span class="nf">extract</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse the output file of hmmer compute from an ordered genes base and produced a new synthetic report file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span>

</div></div>
<div class="viewcode-block" id="OrderedHMMReport"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.OrderedHMMReport">[docs]</a><span class="k">class</span> <span class="nc">OrderedHMMReport</span><span class="p">(</span><span class="n">HMMReport</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    handle HMM report. extract a synthetic report from the raw hmmer output</span>
<span class="sd">    &quot;&quot;&quot;</span>


<div class="viewcode-block" id="OrderedHMMReport.extract"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.OrderedHMMReport.extract">[docs]</a>    <span class="k">def</span> <span class="nf">extract</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parse the output file of hmmer compute from an unordered genes base </span>
<span class="sd">        and produced a new synthetic report file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">i_evalue_sel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">i_evalue_sel</span>
        <span class="n">coverage_treshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cfg</span><span class="o">.</span><span class="n">coverage_profile</span>
        <span class="n">gene_profile_lg</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">profile</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_hmmer_raw_out</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">hmm_out</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">hmm_out</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;&gt;&gt; &quot;</span><span class="p">):</span>
                    <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                    <span class="n">hit_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="n">fields_hit</span> <span class="o">=</span> <span class="n">hit_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;_&#39;</span><span class="p">)</span>
                    <span class="n">replicon_name</span> <span class="o">=</span> <span class="n">fields_hit</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">position_hit</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields_hit</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">10</span>
                    <span class="c"># skip next 2 line</span>
                    <span class="c"># the hit begins on the 3rd line</span>
                    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
                        <span class="n">line</span> <span class="o">=</span> <span class="n">hmm_out</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
                    <span class="k">while</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;  Alignments&quot;</span><span class="p">):</span>
                        <span class="k">print</span> <span class="n">line</span>
                        <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                        <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">5</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="n">i_evalue_sel</span><span class="p">):</span>
                            <span class="n">cov</span> <span class="o">=</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">7</span><span class="p">])</span> <span class="o">-</span> <span class="nb">float</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">gene_profile_lg</span>
                            <span class="k">if</span> <span class="p">(</span><span class="n">cov</span> <span class="o">&gt;=</span> <span class="n">coverage_treshold</span><span class="p">):</span>
                                <span class="n">i_eval</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
                                <span class="n">score</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Hit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">system</span><span class="p">,</span>
                                                     <span class="n">hit_id</span><span class="p">,</span>
                                                     <span class="n">replicon_name</span><span class="p">,</span>
                                                     <span class="n">position_hit</span><span class="p">,</span>
                                                     <span class="n">i_eval</span><span class="p">,</span>
                                                     <span class="n">score</span><span class="p">,</span>
                                                     <span class="n">cov</span><span class="p">))</span>
                        <span class="n">line</span> <span class="o">=</span> <span class="n">hmm_out</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
                                
                                
        <span class="bp">self</span><span class="o">.</span><span class="n">_hmm_extracted</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hits</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

</div></div>
<div class="viewcode-block" id="Hit"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.Hit">[docs]</a><span class="k">class</span> <span class="nc">Hit</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    handle hits found by HMM. the hit are instanciate by :method:`HMMReport.extract` method</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
<div class="viewcode-block" id="Hit.__init__"><a class="viewcode-back" href="../../HMMReport.html#txsscanlib.report.Hit.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">system</span><span class="p">,</span> <span class="n">hit_id</span><span class="p">,</span> <span class="n">replicon_name</span><span class="p">,</span> <span class="n">position_hit</span><span class="p">,</span> <span class="n">i_eval</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">coverage</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        :param gene: the gene corresponding to this profile</span>
<span class="sd">        :type gene: :class:`txsscanlib.gene.Gene` object</span>
<span class="sd">        :param system: the system to which this gene belongs</span>
<span class="sd">        :type system: :class:`txsscanlib.system.System` object</span>
<span class="sd">        :param hit_id: the identifier of the hit</span>
<span class="sd">        :type hit_id: string</span>
<span class="sd">        :param replicon_name: the name of the replicon</span>
<span class="sd">        :type replicon_name: string</span>
<span class="sd">        :param position_hit: the position of the hit on the sequence?</span>
<span class="sd">        :type position_hit: integer</span>
<span class="sd">        :param i_eval: the ???? evalue</span>
<span class="sd">        :type i_eval: float</span>
<span class="sd">        :param score: the score of the hit</span>
<span class="sd">        :type score: float</span>
<span class="sd">        :param coverage: the coverage of the hit</span>
<span class="sd">        :type coverage: float</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">system</span> <span class="o">=</span> <span class="n">system</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">hit_id</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span> <span class="o">=</span> <span class="n">replicon_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">position</span> <span class="o">=</span> <span class="n">position_hit</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">i_eval</span> <span class="o">=</span> <span class="n">i_eval</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">score</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span> <span class="o">=</span> <span class="n">coverage</span>
</div>
    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%d</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\t</span><span class="si">%f</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">replicon_name</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">system</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">i_eval</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
                                                     <span class="bp">self</span><span class="o">.</span><span class="n">coverage</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">__cmp__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">id</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">cmp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">i_eval</span><span class="p">,</span> <span class="n">other</span><span class="o">.</span><span class="n">i_eval</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">cmp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">other</span><span class="o">.</span><span class="n">id</span><span class="p">)</span> </div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../index.html">Txsscan 0.9a1 documentation</a> &raquo;</li>
          <li><a href="../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Sophie Abby, Bertrand Néron.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.2.
    </div>
  </body>
</html>